### PR TITLE
[Merged by Bors] - chore: cache doesn't need to traverse the cache directory

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Cache/IO.lean [hunk 1]
```diff
@@ -209,11 +209,11 @@ abbrev HashMap := Lean.HashMap FilePath UInt64
 
 namespace HashMap
 
-def filter (hashMap : HashMap) (set : Lean.RBTree String compare) (keep : Bool) : HashMap :=
-  hashMap.fold (init := default) fun acc path hash =>
-    let contains := set.contains hash.asLTar
-    let add := if keep then contains else !contains
-    if add then acc.insert path hash else acc
+def filter (hashMap : HashMap) (keep : Bool) : IO HashMap :=
```
Comment (digama0) [inline Cache/IO.lean:212]:
```suggestion
def filterExists (hashMap : HashMap) (keep : Bool) : IO HashMap :=
```
this name was already misleading but now it is even more misleadingm

File: Cache/Requests.lean [hunk 2]
```diff
@@ -60,7 +60,7 @@ def downloadFile (hash : UInt64) : IO Bool := do
 
 /-- Calls `curl` to download files from the server to `CACHEDIR` (`.cache`) -/
 def downloadFiles (hashMap : IO.HashMap) (forceDownload : Bool) (parallel : Bool) : IO Unit := do
-  let hashMap := if forceDownload then hashMap else hashMap.filter (← IO.getLocalCacheSet) false
+  let hashMap := if forceDownload then hashMap else ← hashMap.filter false
```
Comment (digama0) [inline Cache/Requests.lean:63]:
```suggestion
  let hashMap ← if forceDownload then pure hashMap else hashMap.filter false
```
you hit the footgun case of ← lifting over `if`

Comment (semorrison) [reply Cache/Requests.lean:63]: Oof, can we have a linter?

Comment (digama0) [reply Cache/Requests.lean]: This should just be an error in core; most similar cases give you a "can't lift <- over binder" error. I think it's just an oversight.

File: Cache/IO.lean [hunk 3]
```diff
@@ -209,11 +209,11 @@ abbrev HashMap := Lean.HashMap FilePath UInt64
 
 namespace HashMap
 
-def filter (hashMap : HashMap) (set : Lean.RBTree String compare) (keep : Bool) : HashMap :=
-  hashMap.fold (init := default) fun acc path hash =>
-    let contains := set.contains hash.asLTar
-    let add := if keep then contains else !contains
-    if add then acc.insert path hash else acc
+def filterExists (hashMap : HashMap) (keep : Bool) : IO HashMap :=
```
Comment (Vierkantor) [inline Cache/IO.lean:212]: Did I understand the purpose correctly?

```suggestion
/-- Filter the hashmap by whether the entries exist as files in the cache directory.

If `keep` is true, the result will contain the entries that do exist;
if `keep` is false, the result will contain the entries that do not exist.
-/
def filterExists (hashMap : HashMap) (keep : Bool) : IO HashMap :=
```

**Review by Vierkantor — APPROVED** [review]
LGTM otherwise. I'm a bit hesitant to merge this given my Lean 4 metaprogramming knowledge, so:

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**